sycl/doc/SYCL2020-SpecializationConstants.md

```diff
@@ -408,97 +556,217 @@ property_set {
+specialization constants are available, but it will be also used for emulation
+of specialization constants: SPIR-V IDs and sizes of specialization constants
+will be used to calculate offset of each specialization constant within a
+buffer, which is used to propagate them to kernel through kernel arguments.
 
-For each device binary compiler generates a map
-\<Symbolic ID\> =\> \<list of spec constant descriptors\> ("ID map"). DPC++
-runtime imports that map when loading device binaries.
-It also maintains another map \<Spec const symbolic ID\> =\> \<its value\> 
-("value map") per `sycl::kernel_bundle` object. The value map is updated upon
-`kernel_bundler::set_specialization_constant<ID>(val)` and
-`handler::set_specialization_constant<ID>(val)` calls from the app.
+Additionally, another property set will be generated to support emulated
+specialization constants, which will contain a single property with default
+values of all specialization constants in the same form as they will be
+propagated from host to device through kernel arguments, i.e. this property will
+simply contain a blob, which can be used by DPC++ RT to either pre-initialize
+the whole buffer for specialization constants with their default value or to
+extract default value of a particular specialization constant out of it.
 
-In order for runtime to access the right property, it need to compute the
-symbolic ID of a specialization constant based on user-provided inputs, such
-as non-type template argument passed to `set_specialization_constant` argument.
-DPC++ Headers section describes how symbolic IDs are generated and the same
-trick is used within `set_specialization_constant` method:
+For example, the following code:
 ```
-template<auto& SpecName>
-void set_specialization_constant(
-  typename std::remove_reference_t<decltype(SpecName)>::type value) {
-  const char *SymbolicID =
-#if __has_builint(__builtin_unique_stable_name)
-      __builtin_unique_stable_name(detail::specialization_id_name_generator<S>);
-#else
-      // without the builtin we can't get the symbolic ID of the constant
-      "";
-#endif
-  // remember the value of the specialization constant
-  SpecConstantValuesMap[SymbolicID] = value;
+struct Nested {
+  constexpr Nested(float a, float b) : a(a), b(b) {}
+  float a, b;
+};
+struct A {
+  constexpr A(int x, float a, b) : x(x), n(a, b) {}
+  int x;
+  Nested n;
+};
+
+specialization_id<int> id_int(42);
+specialization_id<A> id_A(1, 2.0, 3.0);
+specialization_id<Nested> id_Nested(4.0, 5.0);
+// ...
+  [=](kernel_handler h) {
+    h.get_specialization_constant<id_int>();
+    h.get_specialization_constant<id_A>();
+    h.get_specialization_constant<id_Nested>();
+  }
+```
+
+The following property set will be generated:
+```
+property_set {
+  Name = "SYCL/specialization constants default values",
+  properties: [
+    property {
+      Name: "all",
+      ValAddr: points to byte array [
+        42, // id_int
+        1, 2.0, 3.0, // id_A
```

Comment:
Is it correct to say, that this values are just memory blob representing an object of type `A` constructed using (1, 2.0, 3.0) and not values passed to specialization_id c'tor? 
E.g. for the following:
```

struct C {
   C(int, int, int): Val(42) {}
   int Val = 0;
};

specialization_id<C> id_A(1, 1, 1);
```

The line will be just `42`.

Reply:
> Is it correct to say, that this values are just memory blob representing an object of type A constructed using (1, 2.0, 3.0) and not values passed to specialization_id c'tor?

Yes, that is correct.  I updated a wording and example in 82882b9a91580d741020f2bfe694cf53fab00824 to highlight that